### PR TITLE
[CUDNN] Add an option to force cuDNN usage

### PR DESCRIPTION
**Pull request by eqy**
We've accumulated a variety of magic number shape checks for deciding whether to dispatch to cuDNN (especially e.g., for depthwise convolutions), and these may be out-of-date with newer cuDNN versions, device architectures, etc. This PR adds "force cuDNN" as a way to force dispatching to cuDNN for functionality and perf. testing purposes without rebuilding PyTorch.

cc @csarofeen @ptrblck @xwang233 @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -422,6 +422,9 @@ struct ConvParams {
   // cudnn and miopen are guaranteed not to be on mobile, and T102591915 / T110194934 suggest
   // that maybe the compiledWithCuDNN() check sometimes segfaults (though I can't imagine how)
 #if !defined(C10_MOBILE)
+    if (at::globalContext().userForceCuDNN()) {
```
Comment (mikaylagawarecki) [inline aten/src/ATen/native/Convolution.cpp:425]: Does this mean that we will return True for `use_cudnn` if `force_cudnn` even if `compiledWithCuDNN` is False?

Comment (eqy) [reply aten/src/ATen/native/Convolution.cpp:425]: That's a good point, originally I was thinking that `force_cudnn` should raise any possible failures, but in this case it's probably better to just return a warning about `force_cudnn` not having the intended effect.

**Comment by eqy** [comment]
CC @cpuhrsch

**Comment by cpuhrsch** [comment]
Thanks! Could we also add a few tests?

**Pull request closed without merging**